File: examples/grammar/portfolio-cost/python/_cott_impl/curriculum/portfolio_cost/calculate_portfolio_cost.py
from math import isfinite

from cott_runtime import CottList, Err, F64, Ok, Result
from curriculum.portfolio_cost_types import (
    Holding,
    PortfolioError,
    PortfolioError_NegativePrice,
    PortfolioError_NegativeShares,
    PortfolioError_NonFinitePrice,
    PortfolioError_TotalOverflow,
)

def calculate_portfolio_cost(rows: CottList[Holding]) -> Result[F64, PortfolioError]:
    total = 0.0
    for holding in rows:
        if holding.shares < 0:
            return Err(error=PortfolioError_NegativeShares())
        price = holding.price
        if not isfinite(price):
            return Err(error=PortfolioError_NonFinitePrice())
        if price < 0.0:
            return Err(error=PortfolioError_NegativePrice())
        product = float(holding.shares) * price
        if not isfinite(product):
            return Err(error=PortfolioError_TotalOverflow())
        total = total + product
        if not isfinite(total):
            return Err(error=PortfolioError_TotalOverflow())
    return Ok(value=total)

File: examples/kotlin/grammar/portfolio-cost/kotlin/cott_impl/curriculum/portfolio_cost/calculate_portfolio_cost.kt
package cott_impl.curriculum.portfolio_cost

internal fun calculate_portfolio_cost(rows: cott_runtime.CottList<curriculum.portfolio_cost.Holding>): cott_runtime.CottResult<kotlin.Double, curriculum.portfolio_cost.PortfolioError> {
    var total = 0.0
    for (row in rows) {
        if (row.shares < 0L) {
            return cott_runtime.Err(curriculum.portfolio_cost.PortfolioError.NegativeShares)
        }
        if (!row.price.isFinite()) {
            return cott_runtime.Err(curriculum.portfolio_cost.PortfolioError.NonFinitePrice)
        }
        if (row.price < 0.0) {
            return cott_runtime.Err(curriculum.portfolio_cost.PortfolioError.NegativePrice)
        }
        val product = row.shares.toDouble() * row.price
        if (!product.isFinite()) {
            return cott_runtime.Err(curriculum.portfolio_cost.PortfolioError.TotalOverflow)
        }
        val nextTotal = total + product
        if (!nextTotal.isFinite()) {
            return cott_runtime.Err(curriculum.portfolio_cost.PortfolioError.TotalOverflow)
        }
        total = nextTotal
    }
    return cott_runtime.Ok(total)
}
